Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,311 @@
+by using the following algorithm:
+
+```go
+assert(num_shards >= len(zones))     # Error: zone(s) will not get scraped
+assert(num_shards % len(zones) == 0) # Warning: multi-scraping of instances
+
+shards_per_zone := max(1, floor(num_shards / len(zones)))
+prom_zone_idx := shard_index % shards_per_zone
+    
+if zone_label == zones[prom_zone_idx] {
+    
+    assignment_idx := floor(shard_index / shards_per_zone)
+
+    if hash(address) % shards_per_zone == assignment_idx {
+        do_scrape()
+    }
+}
+```
+
+By using modulo to calculate the `prom_zone_idx`, instances will be distributed
+to zones in the sense of `A,B,C,A,B,C` and so on. This allows a modification of
+the `num_shards` value without redistribution of shards or data.
+This was preferred over allowing the number of `zones` to change, as this is
+less likely to happen.
+
+### Edge cases
+
+We have introduced asserts in the above section to warn about edge cases that
+might lead to duplicate data or data loss.
+
+By the above algorithm. prometheus instances will be distributed in an
+alternating fashion by using the already existing shard index.
+This leads to the following edge cases:
+
+When `num_shards` is 10 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 3. This yields the following distribution
+
+```text
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```text
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
+
+This second case - a zone not being scraped - should lead to an error in the
+operator, causing the change to not be rolled out, as data would be lost.
+The first case - double scraping - should at minimum cause a warning.
+
+## Required configuration options
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus custom resource definition](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus).
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'
+    mode: 'Classic'    
+
+    # Optional: array of relabel configurations to insert before the sharding
+    # relabel configuration.
+    additionalRelabelConfig: [] 
+
+    # The following section is only valid if "mode" is set to "Classic"
+    classic:
+        # Metric label used for sharding.
+        sourceLabel: '__address__'
```

Comment:
I can't think of any ...?

But aside from this, the field _is_ an array, and if we later realize that we _do_ have a case, making this field an array will be a breaking change. What do you think? Allow an array to avoid a potential future breaking change, or keep it simple?

Reply:
I think that the discussion in https://github.com/prometheus-operator/prometheus-operator/pull/7110#discussion_r1843548809 voids the need for a `sourceLabel` field.

Reply:
I disagree (a bit).
The fact that `__address__` is currently hardcoded is not ideal. Allowing users to change this allows for more flexibility and is a (comparable) small change.
As we're currently discussing a change to this section I think it is a good point to add this here, especially as it does not break any existing functionality.

If we had this field now, we could already test a few things in regard to sharding, which we currently cannot.

Reply:
Have you seen that https://github.com/prometheus-operator/prometheus-operator/pull/7126 has merged? The next release of the operator will allow to customize the label(s) on which the hashmod operation is performed.

Reply:
To clarify why I'm relunctant to add new fields: any field that we add we need to live with it until a v2 API happens (if ever) which creates cognitive load for users and maintainers. Not adding it now doesn't mean that it's a permanent no.

Reply:
I personally prefer explicit over implicit, but I get your point.
I'll remove the sourceLabel from both sections and will reference the `__tmp_hash` override.

In case of the `Topology` mode, this is probably a good idea anyways, as the label might be different for pods or services.

Reply:
Added.